cmd/scollector/collectors/snmp_ciscobgp.go

```diff
@@ -0,0 +1,190 @@
+	adminStatus         = ".1.3.6.1.4.1.9.9.187.1.2.5.1.4.1.4"
+	localAS             = ".1.3.6.1.4.1.9.9.187.1.2.5.1.8.1.4"
+	localIdentifier     = ".1.3.6.1.4.1.9.9.187.1.2.5.1.9.1.4"
+	remoteAS            = ".1.3.6.1.4.1.9.9.187.1.2.5.1.11.1.4"
+	remoteIdentifier    = ".1.3.6.1.4.1.9.9.187.1.2.5.1.12.1.4"
+	inUpdates           = ".1.3.6.1.4.1.9.9.187.1.2.5.1.13.1.4"
+	outUpdates          = ".1.3.6.1.4.1.9.9.187.1.2.5.1.14.1.4"
+	inTotalMessages     = ".1.3.6.1.4.1.9.9.187.1.2.5.1.15.1.4"
+	outTotalMessages    = ".1.3.6.1.4.1.9.9.187.1.2.5.1.16.1.4"
+	inUpdateElapsedTime = ".1.3.6.1.4.1.9.9.187.1.2.5.1.27.1.4"
+	acceptedPrefixes    = ".1.3.6.1.4.1.9.9.187.1.2.8.1.1.1.4"
+	deniedPrefixes      = ".1.3.6.1.4.1.9.9.187.1.2.8.1.2.1.4"
+	advertisedPrefixes  = ".1.3.6.1.4.1.9.9.187.1.2.8.1.6.1.4"
+	suppressedPrefixes  = ".1.3.6.1.4.1.9.9.187.1.2.8.1.7.1.4"
+	withdrawnPrefixes   = ".1.3.6.1.4.1.9.9.187.1.2.8.1.8.1.4"
+)
+
+func SNMPCiscoBGP(cfg conf.SNMP) {
+	collectors = append(collectors, &IntervalCollector{
+		F: func() (opentsdb.MultiDataPoint, error) {
+			return c_snmp_ciscobgp(cfg.Community, cfg.Host)
+		},
+		Interval: time.Second * 30,
+		name:     fmt.Sprintf("snmp-ciscobgp-%s", cfg.Host),
+	})
+}
+
+func c_snmp_ciscobgp(community, host string) (opentsdb.MultiDataPoint, error) {
+	// Tag: local_as
+	localASesRaw, err := snmp_ip_tree(host, community, localAS)
+	if err != nil {
+		return nil, err
+	}
+	localASes := make(map[string]string, len(localASesRaw))
+	for k, v := range localASesRaw {
+		s := strconv.FormatInt(v.(int64), 10)
+		if err != nil {
+			return nil, err
+		}
+		localASes[k] = s
+	}
+
+	// Tag: local_id
+	localIdentifiersRaw, err := snmp_ip_tree(host, community, localIdentifier)
+	if err != nil {
+		return nil, err
+	}
+	localIdentifiers := make(map[string]string, len(localIdentifiersRaw))
+	for k, v := range localIdentifiersRaw {
+		localIdentifiers[k] = snmp_combine_ip_uint8(v.([]uint8))
+	}
+
+	// Tag: remote_as
+	remoteASesRaw, err := snmp_ip_tree(host, community, remoteAS)
+	if err != nil {
+		return nil, err
+	}
+	remoteASes := make(map[string]string, len(remoteASesRaw))
+	for k, v := range remoteASesRaw {
+		s := strconv.FormatInt(v.(int64), 10)
+		if err != nil {
+			return nil, err
+		}
+		remoteASes[k] = s
+	}
+
+	// Tag: remote_id
+	remoteIdentifiersRaw, err := snmp_ip_tree(host, community, remoteIdentifier)
+	if err != nil {
+		return nil, err
+	}
+	remoteIdentifiers := make(map[string]string, len(remoteIdentifiersRaw))
+	for k, v := range remoteIdentifiersRaw {
+		remoteIdentifiers[k] = snmp_combine_ip_uint8(v.([]uint8))
+	}
+
+	var md opentsdb.MultiDataPoint
+	// Function to harvest all metrics with the tag groups above
+	add := func(bA bgpAdd) error {
+		m, err := snmp_ip_tree(host, community, bA.oid)
+		if err != nil {
+			return err
+		}
+		for k, v := range m {
+			tags := opentsdb.TagSet{
+				"host":      host,
+				"peer":      k,
+				"local_as":  localASes[k],
```

Comment:
If any of these strings are empty, you will have dataset problems. Do we have a good default if they are empty ("na"?)? Should any of these be metadata rather than tags?